server/routes/posts.js

```diff
@@ -0,0 +1,152 @@
+const express = require("express");
+const router = express.Router();
+const Post = require("../models/Post.js");
+const youtube = require("../youtube.js");
+const github = require("../github.js");
+const qs = require('querystring');
+//get all public posts for All/Explore page
+router.get('/', async (req, res, next) => {
+    console.log(req.body);
+
+    try {
+        const posts = await Post.find({isPublic: true}).exec();
+        res.status(200).send(posts);
+    } catch(err) {
+        res.status(500).send(err);
+    }
+});
+
+//get post by id (pinnedPosts under User)
+router.get('/:id', async (req, res, next) => {
+    try {
+        const id = req.params.id;
+        const post = await Post.findById(id).exec();
+        res.send(post);
+    } catch(err) {
+        res.status(500).send(err);
+    }
+});
+
+//create a new post, will need to get type from user
+router.post('/', async (req, res, next) => {
+    try {
+        if(req.body.url) {
+            const youtubeRegex = /^(https?\:\/\/)?(www\.youtube\.com|youtu\.?be)\/.+$/;
+            const githubRegex = /^(https?\:\/\/)?(www\.github\.com|github\.com)\/.+$/;
+            const url = req.body.url;
+            if(youtubeRegex.test(url)) {
+                const urlToParse = url.split("?")[1];
+                const parsed = qs.parse(urlToParse);
+                const videoId = parsed.v;

```

Comment:
Youtube is weird in that the url comes in all sorts of forms. I don't think there'll always be a `?v=`.

The following example link to the same vid: 
`https://www.youtube.com/watch?v=Wp4EmpeCeE4&feature=youtu.be`
`https://youtu.be/Wp4EmpeCeE4` (link generated from clicking share)

Reply:
For reference: https://stackoverflow.com/questions/3452546/how-do-i-get-the-youtube-video-id-from-a-url